packages/mui-base/src/ModalUnstyled/ModalUnstyled.js

```diff
@@ -242,30 +236,34 @@ const ModalUnstyled = React.forwardRef(function ModalUnstyled(props, ref) {
   }
 
   const Root = components.Root || component;
-  const rootProps = appendOwnerState(
-    Root,
-    {
-      role: 'presentation',
-      ...other,
-      ...componentsProps.root,
+  const rootProps = useSlotProps({
+    elementType: Root,
+    externalSlotProps: componentsProps.root,
+    externalForwardedProps: other,
+    additionalProps: {
       ref: handleRef,
+      role: 'presentation',
       onKeyDown: handleKeyDown,
-      className: clsx(classes.root, componentsProps.root?.className, className),
     },
+    className: classes.root,
     ownerState,
-  );
+  });
 
   const BackdropComponent = components.Backdrop;
-  const backdropProps = appendOwnerState(
-    BackdropComponent,
-    {
+  const backdropProps = useSlotProps({
+    elementType: BackdropComponent,
+    externalSlotProps: componentsProps.backdrop,
+    additionalProps: {
       'aria-hidden': true,
-      open,
       onClick: handleBackdropClick,
-      ...componentsProps.backdrop,
+      open,
     },
     ownerState,
-  );
+  });
+
+  if (!keepMounted && !open && (!hasTransition || exited)) {
```

Comment:
Why was this early exit moved? Isn’t it still better to do this before resolving the props?

Reply:
`useSlotProps` can't be called conditionally as it's a hook.

Reply:
Daah :D